src/libcore/tests/result.rs

```diff
@@ -202,3 +204,30 @@ pub fn test_unwrap_or_default() {
     assert_eq!(op1().unwrap_or_default(), 666);
     assert_eq!(op2().unwrap_or_default(), 0);
 }
+
+#[test]
+fn test_try() {
+    fn try_result_some() -> Option<u8> {
+        let val = Ok(1)?;
+        Some(val)
+    }
+    assert_eq!(try_result_some(), Some(1));
+
+    fn try_result_none() -> Option<u8> {
+        let val = Err(NoneError)?;
+        Some(val)
+    }
+    assert_eq!(try_result_none(), None);
+
+    fn try_result_ok() -> Result<u8, u8> {
+        let val = Ok(1)?;
+        Ok(val)
+    }
+    assert_eq!(try_result_ok(), Ok(1));
```

Comment:
Given that the major concern was the error messages, I agree with @oli-obk that it would be good to add some `src/test/ui` tests covering incorrect usage of `?` when applied to `Option`. E.g., given `x: Option<u32>`:

- `x?` in a function that returns `Result<u32, ()>`,
- `x?` in a function that returns `u32`
- mm anything else to try?

Here is a guide to adding ui tests: https://github.com/rust-lang/rust/blob/master/src/test/COMPILER_TESTS.md#guide-to-the-ui-tests